### ISSUE TITLE
Regression error on 0.27 for client_credentials flow

### ISSUE THREAD

**Issue by AimerLAir** [marcospereirampj/python-keycloak#269]
Hi, it seems that the **client_credentials flow** doesn't work anymore.
The **get_token** method doesn't request token if there is no username and password, what i believe it should do anyway if grant_type is **["client_credentials"]**.

```
def get_token(self):
        token_realm_name = 'master' if self.client_secret_key else self.user_realm_name or self.realm_name
        self.keycloak_openid = KeycloakOpenID(server_url=self.server_url, client_id=self.client_id,
                                              realm_name=token_realm_name, verify=self.verify,
                                              client_secret_key=self.client_secret_key,
                                              custom_headers=self.custom_headers)

        grant_type = ["password"]
        if self.client_secret_key:
            grant_type = ["client_credentials"]
            if self.user_realm_name:
                self.realm_name = self.user_realm_name

        if self.username and self.password:
            self._token = self.keycloak_openid.token(self.username, self.password, grant_type=grant_type)

            headers = {
                'Authorization': 'Bearer ' + self.token.get('access_token'),
                'Content-Type': 'application/json'
            }
        else:
            self._token = None
            headers = {}

        if self.custom_headers is not None:
            # merge custom headers to main headers
            headers.update(self.custom_headers)

        self._connection = ConnectionManager(base_url=self.server_url,
                                             headers=headers,
                                             timeout=60,
                                             verify=self.verify)
```

This behavior was introduced by the #244. (But i do not understand what was its purpose.)

**Comment by j-hellenberg** [comment]
We've also noticed this issue. As you said, currently no token is ever requested for the `client_credentials` flow, which means the library is practically unusable for this authentication method (except by passing some dummy username and password, which will be ignored due to `grant_type` == `["client_credentials"]` but still trigger the if condition...).

Judging by its description, I believe what #244 was trying to achieve was to allow using the library without having to specify any credentials (`username/password` or `client_id/client_secret_key`) at all by directly authenticating using a valid authentication token passed into the `custom_headers` field of the `KeycloakAdmin` constructor (previously, the `get_token` method would throw an error if not provided with any credentials at all, as the call of `self.keycloak_openid.token` would fail in that case). I'm not quite sure why somebody would want to do that though, as the manually provided auth token would eventually expire, at which point the library hasn't got the credential data required to initiate a re-authentication.

Depending on whether the newly introduced use case will be classified as sensible or not, I would vote for either reverting #244 or changing the if-condition to `if (self.username and self.password) or (self.client_id and self.client_secret_key):`.

**Comment by regilero** [comment]
Clearly the case where you do not use username/password but a client_id/secret cannot work here, simply because of this 'if' condition.
Bypassing the if condition by using fake username/password arguments the library now allows for usage of service accounts with client secret (and not only on master realm like with version 0.26). I vote for a rapid hotfix on that if condition.

**Comment by andricDu** [comment]
I've submitted a PR with a fix for this: https://github.com/marcospereirampj/python-keycloak/pull/281

**Comment by TomasChmelik** [comment]
Any news on this? I'd like to use client_credentials flow to get token for non-master realm. So I can't use neither 0.26.1 nor 0.27.0

**Comment by dominiks96** [comment]
@TomasChmelik A workaround for this issue is to set username and password to some dummy value, that will never be used. In the current code, your client credential will then be properly used.

**Comment by TomasChmelik** [comment]
That workaround doesn't work, because [this](https://github.com/marcospereirampj/python-keycloak/blob/859ea999481213cf6adf7718e56b44ee599c371a/keycloak/keycloak_openid.py#L173) doesn't use client secret key so it will fail to authenticate.

**Comment by dominiks96** [comment]
@TomasChmelik yes it does. https://github.com/marcospereirampj/python-keycloak/blob/859ea999481213cf6adf7718e56b44ee599c371a/keycloak/keycloak_openid.py#L200

**Comment by TomasChmelik** [comment]
Then I don't know why it didn't work, I just know it didn't. Maybe some config issue with my cluster

**Comment by ryshoooo** [comment]
Closing the issue as the fix was merged in #324
